[P020 Ser2Net] Serial port selection on ESP32

TD-er opened the pull request:
See: https://github.com/letscontrolit/ESPEasy/issues/2750#issuecomment-770416257
Made by @immiimmi

TD-er committed:
[P020 Ser2Net] Serial port selection on ESP32

See: https://github.com/letscontrolit/ESPEasy/issues/2750#issuecomment-770416257
Made by @immiimmi

TD-er commented:
Breaking changes:
- Specific UART settings are not included in this change (e.g. start/stop bits)
- Command `serialsend` is commented out.

Not sure yet if these are the only breaking changes.
Would be nice if those could be "unbroken" before being merged.

immiimmi commented:
Thanks for considering my proposal, and reviewing it
->Specific UART settings are not included in this change (e.g. start/stop bits)
I did not understand your comment; I saved it in #define P020_SERIAL_CONFIG;
see the image
->Command serialsend is commented out.
I did not really get the function, therefore I comment it out.
If you could describe the purpose of the serialsend inside ser2net, I can implement it.
immi

<img width="627" alt="2021-01-31-21 37 06" src="https://user-images.githubusercontent.com/11799084/106397476-f2746980-640d-11eb-889c-7cf62d73cc1c.png">

TD-er commented:
Ah OK, I just looked over it briefly and missed the serial port config.
Maybe I should not review code on the couch when watching a TV show, but rather build and execute it :)

Nauttor commented:
Hi @TD-er and @immiimmi,

I've been waiting for this fix in order to be able to use the esp32 as ser2net. I tried to build it myself but I have errors in my build environment to resolve. Meanwhile may you share with me the bin to test the commented fix?  
Thanks in advance and keep the good work.

TD-er committed:
Merge branch 'mega' into bugfix/P020_ser2net

TD-er commented:
@Nauttor Just started a test build.
Will be ready in +/- 30 minutes, but not sure if I can then also upload it right away.

Nauttor commented:
> @Nauttor Just started a test build.
> Will be ready in +/- 30 minutes, but not sure if I can then also upload it right away.

Great, thanks.

TD-er commented:
A test build:
- [ESPEasy_ESP82xx_mega-20210114-53-PR_3483.zip](https://www.dropbox.com/s/xhadhmywfi62znb/ESPEasy_ESP82xx_mega-20210114-53-PR_3483.zip?dl=0)
- [ESPEasy_ESP32_mega-20210114-53-PR_3483.zip](https://www.dropbox.com/s/khex70nk6k3u3uj/ESPEasy_ESP32_mega-20210114-53-PR_3483.zip?dl=0)

Nauttor commented:
> A test build:
> 
> * [ESPEasy_ESP82xx_mega-20210114-53-PR_3483.zip](https://www.dropbox.com/s/xhadhmywfi62znb/ESPEasy_ESP82xx_mega-20210114-53-PR_3483.zip?dl=0)
> * [ESPEasy_ESP32_mega-20210114-53-PR_3483.zip](https://www.dropbox.com/s/khex70nk6k3u3uj/ESPEasy_ESP32_mega-20210114-53-PR_3483.zip?dl=0)

HI @TD-er ,

I load the firmware you posted, but the interface that I get is different from the one in the screenshot from @immiimmi. It appears as the previous one without the "serial config" label. I don't know if it's expected or not, can you confirm that? 
<img width="786" alt="Screenshot 2021-02-04 at 19 09 12" src="https://user-images.githubusercontent.com/17730164/106936226-9ca90580-671c-11eb-81d0-0e29b3b6852d.png">

TD-er commented:
I will try the build later this evening myself.
I have not tested it myself, I only used the file in the ZIP file attached on the forum post.

Nauttor commented:
> I will try the build later this evening myself.
> I have not tested it myself, I only used the file in the ZIP file attached on the forum post.

I try the ser2net service and it crash badly the esp32 as occurred before in the previous build.  Every time I click submit to a simple conf like the one on the screenshot and then reboot it never comes back from the dead. It needs to be re-flashed. It's the same behaviour I experience before. I tried to disable the serial in advanced too and nothing changes, it crash every time I applied a conf in devices after reboot. The device boot like this:


INIT : Booting version:  (ESP32 SDK v3.2.3-14-gd3e562907)
31 : Info   : INIT : Free RAM:291052
31 : Info   : INIT : Cold Bootets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DOUT, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:1044
load:0x40078000,len:8896
load:0x40080400,len:5816
entry 0x400806ac
⸮U30 : Info   : 

<img width="753" alt="Screenshot 2021-02-04 at 19 41 57" src="https://user-images.githubusercontent.com/17730164/106942583-a6366b80-6724-11eb-85a8-73de0b32bf4a.png">

BTW the hardware is a ESP-WROOM-32

immiimmi commented:
Dear @Nauttor, 
considering your screenshot, you do not have my implemetation of ser2net running.
In my version you can chose the serial port (see my post above) https://github.com/letscontrolit/ESPEasy/pull/3483#issuecomment-770448500

Nauttor commented:
> Dear @Nauttor,
> considering your screenshot, you do not have my implemetation of ser2net running.
> In my version you can chose the serial port (see my post above) [#3483 (comment)](https://github.com/letscontrolit/ESPEasy/pull/3483#issuecomment-770448500)

Hi @immiimmi, 

That's what I told @TD-er , as I didn't see the same interface my first thought was that the file he upload is not well build witht your plugin or something is incomplete or missing in the build.

Do you a binary where I can test the ser2net?

TD-er commented:
Can you check the build timestamp on the sysinfo page?
Just to be sure you actually flashed a new build.

I just checked the source code as checked out on my build machine and that contains the changes from this PR.

Nauttor commented:
> Can you check the build timestamp on the sysinfo page?
> Just to be sure you actually flashed a new build.
> 
> I just checked the source code as checked out on my build machine and that contains the changes from this PR.

You are absolutely right, I misspelt the name because the date was the same in the name of the file and flashed the wrong version. MY BAD, sorry. Testing the new version now, I'll report back. Thanks

Nauttor commented:
Hi @TD-er and @immiimmi ,

After loading the correct file it show the interface as expected and I'm testing it with my Opentherm Gateway serial output. 

@TD-er,  I think I find a bug  in the Wifi setup. 
My network have 2 SSID, lets say SSIDA, SSIDB. SSIDA have WPA2Enterprise, an SSIDB WPA2/PSK. 
In the log it looks like:

836784: WIFI : Scan result: #Hidden# B4:XX:XX:XX:XX:76 Ch:6 (-58dBm) WPA2 Enterprise
836784: WIFI : Scan result: #Hidden# B6:XX:XX:XX:XX:76 Ch:6 (-58dBm) WPA2/PSK

I setup in the dialog the ESP32 to connect to the SSIDB B6:FB:E4:9A:84:76 Ch:6 (-58dBm) WPA2/PSK, However look at the following lines about what the ESP try to do:

836784: WIFI : Scan result: #Hidden# B4:XX:XX:XX:XX:76 Ch:6 (-58dBm) WPA2 Enterprise
836784: WIFI : Scan result: #Hidden# B6:XX:XX:XX:XX:76 Ch:6 (-58dBm) WPA2/PSK
836784: WIFI : Credentials Changed, retry connection. SSID: SSIDB
836785: WIFI : Connecting SSIDB#Hidden#  B4:XX:XX:XX:XX:76 Ch:6 (-58dBm) WPA2 Enterprise attempt #0
836793: WIFI : Arduino wifi status: WL_NO_SSID_AVAIL ESPeasy internal wifi status: DISCONNECTED

It tries to connect to the BSSID B4:XX:XX:XX:XX:76 which is the enterprise network, not the one it should. I'm not transmitting the SSIDB from the BSSID that start with B4. It's nonsense. 

And just to add more annoying data to this behaviour, I tried 2 module, 1 ESP-WROOM-32 and two ESP32-S, It work as expected in the WROOM and not in boths ESP-32S. I tried to reflash both several times, always the same result.

TD-er commented:
Those ESP32-S units, did they work with previous builds of ESPEasy?
The WiFi changes have not yet been present in previous builds on the releases page.

I'm not sure if the ESP can even connect to the WPA2 Enterprise. If not, then I can leave it out of the attempts to connect.
For more info on the recent changes in WiFi, see https://github.com/letscontrolit/ESPEasy/pull/3473#issuecomment-766366910

Nauttor commented:
> Those ESP32-S units, did they work with previous builds of ESPEasy?
> The WiFi changes have not yet been present in previous builds on the releases page.
> 
> I'm not sure if the ESP can even connect to the WPA2 Enterprise. If not, then I can leave it out of the attempts to connect.
> For more info on the recent changes in WiFi, see [#3473 (comment)](https://github.com/letscontrolit/ESPEasy/pull/3473#issuecomment-766366910)

@TD-er  you are absolutely right, the latest public release from January connect to the wifi as expected and did not replicate this strange behaviour. I did the test flashing the latest from January and it connects without any issue. If I load the one you post here it fail. And I did one more test, I load the one from January, I setup the wifi successfully and then I upgrade it from the web page with the latest bin you upload, it failed too. There must be something in the new code related to the wifi that broke something.

Nauttor commented:
@TD-er , About if ESP can connect to WPA enterprise it can, not Esplanade's easy as it is not implemented as I understand, but you have an example here for ESP32, [https://gist.github.com/Matheus-Garbelini/332dc35dbdf640d4e5f2672636e19bde](https://gist.github.com/Matheus-Garbelini/332dc35dbdf640d4e5f2672636e19bde). 

I read the PR for the hidden Networks, looking at my previous log, it appears to me as if the ESP ignore the first 2 digits B4 and B6 in the BSSID and just compare the rest and because there are identical it tries to connect to the WPA Enterprise. Can this be possible? Maybe is not looking all the bits from the BSSID. Even if that's the case it does not explain why the WROOM works and the ESP-32S not.

836784: WIFI : Scan result: #Hidden# B4:XX:XX:XX:XX:76 Ch:6 (-58dBm) WPA2 Enterprise
836784: WIFI : Scan result: #Hidden# B6:XX:XX:XX:XX:76 Ch:6 (-58dBm) WPA2/PSK
836784: WIFI : Credentials Changed, retry connection. SSID: SSIDB
836785: WIFI : Connecting SSIDB#Hidden# B4:XX:XX:XX:XX:76 Ch:6 (-58dBm) WPA2 Enterprise attempt #0

TD-er commented:
The procedure is as follows:
- Scan for WiFi networks
- Pick out scan results matching known SSID credentials and add them to a list of candidates.
- For hidden SSIDs, create a candidate for each combination of known SSID and the found MAC address
- Add current active connection to the candidates list with a "fake" RSSI of -1 to always be the strongest
- Sort list of candidates, strongest RSSI first.
- Process all items on the list until list is empty, then scan again to create a new list

There is one exception to this. If the current active connection is considered to be stable (after N minutes), the list is cleared and the current active connection is added to the (now empty) list of candidates.

If you know you don't need to connect to a hidden SSID, you can disable it (I think you need to explicitly enable it to connect to hidden SSID, maybe that's what's wrong with your ESP32S units?

Problem with hidden SSIDs is that you don't know what SSID will match.
So you have to try them all.
But as you may have noticed, an attempt does fail quite fast, so not a lot of time is lost here and we can assume your APs may have a stronger signal compared to those of neighbors, so your own APs will probably be tested first.


Do you have a log of the ESP32S units that fail to connect?

Nauttor commented:
> The procedure is as follows:
> 
>     * Scan for WiFi networks
> 
>     * Pick out scan results matching known SSID credentials and add them to a list of candidates.
> 
>     * For hidden SSIDs, create a candidate for each combination of known SSID and the found MAC address
> 
>     * Add current active connection to the candidates list with a "fake" RSSI of -1 to always be the strongest
> 
>     * Sort list of candidates, strongest RSSI first.
> 
>     * Process all items on the list until list is empty, then scan again to create a new list
> 
> 
> There is one exception to this. If the current active connection is considered to be stable (after N minutes), the list is cleared and the current active connection is added to the (now empty) list of candidates.
> 
> If you know you don't need to connect to a hidden SSID, you can disable it (I think you need to explicitly enable it to connect to hidden SSID, maybe that's what's wrong with your ESP32S units?
> 
> Problem with hidden SSIDs is that you don't know what SSID will match.
> So you have to try them all.
> But as you may have noticed, an attempt does fail quite fast, so not a lot of time is lost here and we can assume your APs may have a stronger signal compared to those of neighbors, so your own APs will probably be tested first.
> 
> Do you have a log of the ESP32S units that fail to connect?

I reproduce the exact same setup in both hardware WROOM and ESP32-s, enabling the Hidden SSID CHECKBOX, and WROOM works the others not. The previous version works in both hardwares, so something in the PR should have affect them. 

I'll try to reproduce the connection in both hardwares and extract the logs to report them back.

TD-er commented:
Problem is, I don't have an ESP32S to test.
Also I can't find it on [Espressif's site](https://www.espressif.com/en/products/modules/esp32)

Nauttor commented:
> Problem is, I don't have an ESP32S to test.
> Also I can't find it on [Espressif's site](https://www.espressif.com/en/products/modules/esp32)

You can find it as a quite popular module named esp32cam, and nodemcu32s.
I have 2 esp32cam, it's quite usefull because it have an external connection for antenna. Count with me for whatever tests you need. I'll post back later with the wifi logs. 

![Ai-Thinker-NodeMCU-32S-ESP32-Development-Board2](https://user-images.githubusercontent.com/17730164/106972952-2626fa80-6752-11eb-9cdc-3ecb5fefd4e5.png)
![ce798ebf-4fc6-42f8-9386-c8823e54f943](https://user-images.githubusercontent.com/17730164/106972956-26bf9100-6752-11eb-8222-482a1a1d04c8.jpg)

immiimmi commented:
> After loading the correct file it show the interface as expected and I'm testing it with my Opentherm Gateway serial output.

Dear  @Nauttor , if I understand correctly this pull request is working for you. I would  suggest that your wifi problem is unrelated and should be discussed in a separated issue.

Dear @TD-er , how can I help you to get the new ser2net merged? There is still the "Command serialsend is commented out"
I still do not know the requirements related to serialsend, therefore I comment it out. If you could describe the purpose of the serialsend inside ser2net, I can implement it. 

b.r.
immi

uzi18 commented:
serialsend should send data from argument via serial, but maybe we should have also opposite command like netsend to send data to client

Nauttor commented:
> > After loading the correct file it show the interface as expected and I'm testing it with my Opentherm Gateway serial output.
> 
> Dear @Nauttor , if I understand correctly this pull request is working for you. I would suggest that your wifi problem is unrelated and should be discussed in a separated issue.
> 
> Dear @TD-er , how can I help you to get the new ser2net merged? There is still the "Command serialsend is commented out"
> I still do not know the requirements related to serialsend, therefore I comment it out. If you could describe the purpose of the serialsend inside ser2net, I can implement it.
> 
> b.r.
> immi

Hi @TD-er ,

I do some fresh test this morning and I get to the conclusion that the Hidden network wifi issue appears to be related to my infrastructure, if anything resurface I will report it back in the other topic. So please forget my previous assumptions about the wifi. 

About the Ser2Net It appears to be working as expected. If I encounter something on it I'll report back. 

Thanks

TD-er commented:
> serialsend should send data from argument via serial, but maybe we should have also opposite command like netsend to send data to client

My objection to both "serialsend" and "netsend" is that they are quite generic which is something I would like to prevent in plugin commands and rather expect on internal commands.

uzi18 commented:
Just mention idea behind these commands, you cannot send anything when your client is not connected or ser2net configured, after merging @tonhuisman changes it will be possible to address specific task or maybe it is already merged ;)
By default it could use first ser2net task.
Maybe if it is not documented anywhere, we can choose different command name

immiimmi commented:
dear @TD-er and dear @uzi18 
I cleaned up the 3 files and implemented the 2 commands
**ser2netclientsend** which sends to client  
**serialsend**  which sends to serial port (it is documented here https://www.letscontrolit.com/wiki/index.php/ESPEasy_Command_Reference)

I do not need the 2 commands, but I can live with them :)
[P20_immi_with_serialsend.zip](https://github.com/letscontrolit/ESPEasy/files/5934785/P20_immi_with_serialsend.zip)

immiimmi commented:
Dear @TD-er,
could you find some time to review my last zip file https://github.com/letscontrolit/ESPEasy/pull/3483#issuecomment-774238886 ?
I think it addersses all your perplexities.
If you have more wishes/hints, I can try to fullfill them.

p.s. I really like the espeasy concept and I think providing 3 ser2net (working simultanuesly out of the box) is pretty unique.
e.g. I am using it for two energy meters and one heatpump, sending the serial data to fhem. I still have free pins for measuring temperature/humidity in the basement.

b.r 
immi

TD-er commented:
I will later this evening.

TD-er committed:
Merge branch 'mega' into bugfix/P020_ser2net

TD-er committed:
[P020 Ser2Net] Restore serialsend command & add ser2netclientsend

Thanks @immiimmi

TD-er commented:
Took a bit longer, but I processed your changes.
Will make a test build, so people can test it.

TD-er commented:
Made a test build:
- [ESPEasy_ESP32_mega-20210223-34-PR_3483.zip](https://www.dropbox.com/s/s92829snd21jbcp/ESPEasy_ESP32_mega-20210223-34-PR_3483.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20210223-34-PR_3483.zip](https://www.dropbox.com/s/5erzlx9uj1r91d4/ESPEasy_ESP82xx_mega-20210223-34-PR_3483.zip?dl=0)

immiimmi commented:
In my case, I know it works considering I wrote it :)
It could be usefull an indipendent  test
dear @Nauttor, would you mind testing the last testbuild from @TD-er?

elektron-bbs commented:
Excuse my bad English.
I use custom build on an ESP32 and get this message with plugin 020:

`PLUGIN_USES_SERIAL not defined`

I think the following line should be in the define_plugin_sets.h file
```
#if defined(USES_P049) || defined(USES_P052) || defined(USES_P053) || defined(USES_P056) || defined(USES_P071) || defined(USES_P075) || defined(USES_P078) || defined(USES_P082) || defined(USES_P085) || defined(USES_P087) || defined(USES_P094) || defined(USES_P102) || defined(USES_P108) || defined(USES_C018)
```
to be added to
```
#if defined(USES_P020) || defined(USES_P049) || defined(USES_P052) || defined(USES_P053) || defined(USES_P056) || defined(USES_P071) || defined(USES_P075) || defined(USES_P078) || defined(USES_P082) || defined(USES_P085) || defined(USES_P087) || defined(USES_P094) || defined(USES_P102) || defined(USES_P108) || defined(USES_C018)
```

TD-er commented:
Yep, that should be patched too.
Thanks for reporting it.

TD-er committed:
Merge branch 'mega' into bugfix/P020_ser2net

TD-er committed:
[Build] Make sure PLUGIN_USES_SERIAL is defined for USES_P020

As reported here: https://github.com/letscontrolit/ESPEasy/pull/3483#issuecomment-796036102

TD-er committed:
Merge branch 'mega' into bugfix/P020_ser2net

TD-er committed:
[P020 Ser2Net] Re-apply connected event (#3587)

TD-er commented:
@uzi18 I re-applied your PR to this pending PR.
Can you test if it is working, then I can merge this.

HomeAutoUser commented:
Hello @TD-er 
I found your branch and flashed it to my hardware. Top.
So far I have not noticed any errors.

TD-er commented:
> Hello @TD-er
> I found your branch and flashed it to my hardware. Top.
> So far I have not noticed any errors.

What serial settings are you using?
Anything other than the default Serial0 port?

elektron-bbs commented:
I have coupled a SIGNALduino (https://github.com/RFD-FHEM/SIGNALDuino) to ESPEasy (ESP32) via Telnet with FHEM (https://fhem.de/). The connection is terminated approximately every 12032 seconds. Is that a mistake or something intentional?

```
Message                                                                                        | Time     | Diff (Seconds)
-----------------------------------------------------------------------------------------------|----------|---------------
2021.04.22 01:48:38 1: 192.168.178.44:23 disconnected, waiting to reappear (sduinoEasyPico434) | 01:48:38 |	
2021.04.22 05:09:10 1: 192.168.178.44:23 disconnected, waiting to reappear (sduinoEasyPico434) | 05:09:10 | 12032
2021.04.22 08:29:43 1: 192.168.178.44:23 disconnected, waiting to reappear (sduinoEasyPico434) | 08:29:43 | 12033
2021.04.22 11:50:15 1: 192.168.178.44:23 disconnected, waiting to reappear (sduinoEasyPico434) | 11:50:15 | 12032
2021.04.22 15:10:48 1: 192.168.178.44:23 disconnected, waiting to reappear (sduinoEasyPico434) | 15:10:48 | 12033
2021.04.22 18:31:20 1: 192.168.178.44:23 disconnected, waiting to reappear (sduinoEasyPico434) | 18:31:20 | 12032
2021.04.22 21:51:53 1: 192.168.178.44:23 disconnected, waiting to reappear (sduinoEasyPico434) | 21:51:53 | 12033
```

TD-er commented:
It sure looks like a regular interval on your setup, but doesn't look to me as logical interval. (200 minutes, 33 sec)

What connection is this?
MQTT?

Is the ESP rebooting?
Do you see the WiFi reconnect counter increase too?

TD-er commented:
Just one question as you apparently build it yourself.
Can you build it based on the latest code in the mega branch?
And is "Periodical WiFi scan" checked? (tools -> Advanced, bottom of the page)

elektron-bbs commented:
> What connection is this?
> MQTT?

Only FHEM HTTP is set up as a controller. But that shouldn't have anything to do with ser2net.
 
> Is the ESP rebooting?

No, uptime is over 6 days.

> Do you see the WiFi reconnect counter increase too?

I didn't pay attention to that, but there are no entries in my router's logs.

TD-er commented:
See the sysinfo page in ESPEasy.

elektron-bbs commented:
> Can you build it based on the latest code in the mega branch?

The build is based on 20111:

![image](https://user-images.githubusercontent.com/26026988/115917623-d0bbd900-a476-11eb-9700-2a79f661c619.png)

When i build with 20112 the load is over 30%. With 20111 is under 10%.

> And is "Periodical WiFi scan" checked? (tools -> Advanced, bottom of the page)

I don't find this attitude there.

![image](https://user-images.githubusercontent.com/26026988/115918429-cd751d00-a477-11eb-8ea1-4d8310b71ed3.png)

TD-er commented:
The "20111" is not a build version, but rather a settings file version.
It is a bit confusing I know.
This code was not yet merged at that date, so I honestly have no idea what code your build was based on.
Apparently the "Periodical Scan WiFi" was not yet merged in this build, or else it would have been a setting on the Advanced settings page.

Yesterday I made a lot of improvements regarding checking the success of memory allocations, so that could perhaps fix these issues you're seeing.
It is also possible your system has some periodical actions (tasks) running which show some drift in when they are being processed which come together after exactly 200 minutes + some 30-seconds loop which put some extra load on the system which causes it to run out of resources. (then yesterday's fixes might work)

elektron-bbs commented:
i will try tomorrow.

elektron-bbs commented:
I downloaded the latest sources today. This creates a new problem: The P012 LCD2004 plug-in only partially works. Texts directly in the plugin are displayed. Texts that are transferred using rules or via HTTP are no longer displayed.

TD-er commented:
Can you give an example of the exact rules line and/or HTTP call?
And do you have a password set on your node? (should not make a difference for rules processing though)

elektron-bbs commented:
Yes, in rules:

```
on Clock#Time=All,**:** do // will run once every minute
 if [DummyWarn#Gefrierschrank]=0 and [DummyWarn#Kuehlschrank]=0 and [DummyWarn#Dummy_3]=0 and [DummyWarn#Dummy_4]=0 
  GPIO,33,0 // LED rot aus
  if %sysyear%>2020
   LCD,1,1,"%sysday_0%.%sysmonth_0%.%sysyear% %systm_hm%"
  endif
 endif
.......
```

Log:

```
85429: EVENT: Clock#Time=Sat,15:24
85533: ACT : GPIO,33,0
85534: GPIO : port#33: set to 0
85547: ACT : LCD,1,1,'24.04.2021 15:24'
```

Password is not set.

TD-er commented:
Hmm that plugin has not changed in months.
What was the previous build that worked like this?

TD-er commented:
Can you make a separate issue for this display problem?
We're now discussing it at an already merged PR, so hard to find again and also others will probably not see it.

TD-er commented:
Oh and maybe just for testing , try to change the dots in the data into something else like an underscore to make sure it isn't parsed as a floating point value.

elektron-bbs commented:
Release mega-20210223 works.
I am creating a new issue.
The points are not in the rules. I only added them here to indicate that I shortened a bit there.

uzi18 commented on src/src/PluginStructs/P020_data_struct.cpp:
```diff
@@ -78,6 +78,7 @@ bool P020_Task::hasClientConnected() {
     if (ser2netClient) { ser2netClient.stop(); }
     ser2netClient = ser2netServer->available();
     ser2netClient.setTimeout(CONTROLLER_CLIENTTIMEOUT_DFLT);
+    sendConnectedEvent(false);
```
@TD-er true here

TD-er replied:
Thanks :)
